llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,230 @@
+    // duck module verifier complaints.
+    Type *ParamType = CASPrintf->getFunctionType()->getParamType(0);
+    if (Const->getType() != ParamType)
+      Const = ConstantExpr::getPointerCast(Const, ParamType);
+    CI->setArgOperand(0, Const);
+  }
+
+private:
+  CallInst *CI;
+  Constant *CASArg;
+};
+
+/// The function's effect is similar to V->stripPointerCastsAndAliases(), but
+/// also strips load/store aliases.
+Value *stripToMemorySource(Value *V) {
+  Value *MemoryAccess = V;
+  if (auto *LI = dyn_cast<LoadInst>(MemoryAccess)) {
+    Value *LoadSource = LI->getPointerOperand();
+    auto *Store = cast<StoreInst>(*llvm::find_if(
+        LoadSource->users(), [](User *U) { return isa<StoreInst>(U); }));
+    MemoryAccess = Store->getValueOperand();
+  }
+  return MemoryAccess->stripPointerCastsAndAliases();
+}
+
+void AddrspaceReplacer::runOnFunctionDeclaration(Function *F) {
+  auto *LiteralType = F->getArg(0)->getType();
+  if (LiteralType->getPointerAddressSpace() == ConstantAddrspaceID)
+    // No need to replace the literal type and its printf users
+    return;
+  SmallVector<CallReplacer, 16> CallsToReplace;
+  for (User *U : F->users()) {
+    if (!isa<CallInst>(U))
+      continue;
+    auto *CI = cast<CallInst>(U);
+
+    /// This key algorithm reaches the global string used as an argument to a
+    /// __spirv_ocl_printf call. It then generates a constant AS copy of that
+    /// global (or gets an existing one). For the return value, the call
+    /// instruction is paired with its future constant addrspace string
+    /// argument.
+    Value *Stripped = stripToMemorySource(CI->getArgOperand(0));
+    if (auto *Literal = dyn_cast<GlobalVariable>(Stripped))
+      CallsToReplace.emplace_back(CI, getCASLiteral(Literal));
+    else if (auto *Arg = dyn_cast<Argument>(Stripped)) {
+      // The global literal is passed to __spirv_ocl_printf via a wrapper
+      // function argument. We'll update the wrapper calls to use the builtin
+      // function directly instead.
+      Function *WrapperFunc = Arg->getParent();
+      for (User *WrapperU : WrapperFunc->users()) {
+        auto *WrapperCI = cast<CallInst>(WrapperU);
+        Value *StrippedArg = stripToMemorySource(WrapperCI->getArgOperand(0));
+        // We're only expecting 1 level of wrappers, so cast unconditionally
+        auto *Literal = cast<GlobalVariable>(StrippedArg);
+        CallsToReplace.emplace_back(WrapperCI, getCASLiteral(Literal));
+      }
+      // We're certain that the wrapper won't have any uses, since we've just
+      // marked all its calls for replacement with __spirv_ocl_printf.
+      FunctionsToDrop.emplace_back(WrapperFunc);
+      // Similar certainty for the generic AS version of __spirv_ocl_printf
+      // itself - we've determined it only gets called inside the
+      // soon-to-be-removed wrapper.
+      assert(F->hasOneUse() && "Unexpected __spirv_ocl_printf call outside of "
+                               "SYCL wrapper function");
```

Comment:
Once the wrapper name check is done as per your [above suggestion](https://github.com/intel/llvm/pull/5069/files#r769858239), this scenario should be excluded.

Reply:
Ok, but what if there are three user-defined functions? It just seems weird to me that we are asserting amount of uses of a function while iterating over them in range-based loop. There could be 10 users, but right after handling the first one we by some reason expect the rest nine disappear?

Reply:
Since the block is aimed at the case where `__spirv_ocl_printf` calls are made inside the non-inlined `experimental::printf`, we're expecting these "10 users" to call `experimental::printf` instead. These 10 users have already been marked for replacement on lines 213-223, so we can safely remove the generic AS `__spirv_ocl_printf` version once the users have been visited.
Moreover, it's all the same with the variadic and non-variadic cases - variadic `__spirv_ocl_printf(const char *, ...)` will always be called through `experimental::printf(const char *, ...), non-variadic `__spirv_ocl_printf(const char *, type_specifier value)` will always be reached through `experimental::printf(const char *, type_specifier value)`.

Reply:
And to clarify further, it's the builtin function (`F`) whose use count is asserted upon, not the `WrapperF` (`experimental::printf`) which has already been marked for removal at the previous code line.